### PR TITLE
fix: add VTIMEZONE to Appointments

### PR DESCRIPTION
**Pull request by miaulalala**
fixes part of https://github.com/nextcloud/calendar/issues/5031

Tested it but this doesn't fix the diverging timezones. Still, it makes us RFC compliant which is always a good thing, and might fix Outlook, although I don't have an Outlook setup to test.

Before:

```
BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Sabre//Sabre VObject 4.5.3//EN
CALSCALE:GREGORIAN
BEGIN:VEVENT
UID:sabre-vobject-4d193331-955f-4041-a7b9-44b14809b030
DTSTAMP:20230914T191309Z
SUMMARY:Bob - Have Lunch With Me
STATUS:CONFIRMED
DTSTART;TZID=Europe/Vienna:20230918T120000
DTEND;TZID=Europe/Vienna:20230918T123000
DESCRIPTION:Testing
ORGANIZER;CN=Anna Admin;CUTYPE=INDIVIDUAL;PARTSTAT=ACCEPTED:mailto:xxxxx
 xx@gmx.net
ATTENDEE;CN=Anna Admin;CUTYPE=INDIVIDUAL;RSVP=TRUE;ROLE=REQ-PARTICIPANT;PAR
 TSTAT=ACCEPTED:mailto:xxxxxxxx@gmx.net
ATTENDEE;CN=Bob;CUTYPE=INDIVIDUAL;RSVP=TRUE;ROLE=REQ-PARTICIPANT;PARTSTAT=A
 CCEPTED:mailto:xxxxxxx@gmail.com
LOCATION:Abigail's Kitchen
X-NC-APPOINTMENT:iHFJtxxcrW8g
END:VEVENT
END:VCALENDAR
```

After:

```
BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Sabre//Sabre VObject 4.5.3//EN
CALSCALE:GREGORIAN
BEGIN:VTIMEZONE
TZID:Europe/Vienna
BEGIN:DAYLIGHT
DTSTART:20230326T010000
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
END:DAYLIGHT
BEGIN:DAYLIGHT
DTSTART:20240331T010000
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
END:DAYLIGHT
BEGIN:STANDARD
DTSTART:20231029T010000
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
TZNAME:CET
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
UID:sabre-vobject-27752eed-efcb-4972-8cd5-84f78b65c6ae
DTSTAMP:20231102T134259Z
SUMMARY:Test - Have Lunch With Me
STATUS:CONFIRMED
DTSTART;TZID=Europe/Vienna:20231106T120000
DTEND;TZID=Europe/Vienna:20231106T123000
DESCRIPTION:Testing
ORGANIZER;CN=Anna Admin;CUTYPE=INDIVIDUAL;PARTSTAT=ACCEPTED:mailto:xxxxxxxxx@gmx.net
ATTENDEE;CN=Anna Admin;CUTYPE=INDIVIDUAL;RSVP=TRUE;ROLE=REQ-PARTICIPANT;PAR
 TSTAT=ACCEPTED:mailto:xxxxxxxx@gmx.net
ATTENDEE;CN=Test;CUTYPE=INDIVIDUAL;RSVP=TRUE;ROLE=REQ-PARTICIPANT;PARTSTAT=
 ACCEPTED:mailto:xxxxxxxx@gmail.com
LOCATION:Abigail's Kitchen
X-NC-APPOINTMENT:iHFJtxxcrW8g
END:VEVENT
END:VCALENDAR
```

### CODE REVIEW

**Review by ChristophWurst — COMMENTED** [review]
Is it sufficient to only have TZID or do we need the full tz spec?

**Comment by ChristophWurst** [comment]
unassigning milestone because this will go into main

**Comment by miaulalala** [comment]
> Is it sufficient to only have TZID or do we need the full tz spec?

I need to look into the sabre code more but having the full timezone info could be good to have.

**Comment by miaulalala** [comment]
With https://gist.github.com/thomascube/47ff7d530244c669825736b10877a200 it works fine, for timezones that have offests it also works well.

File: lib/Service/Appointments/BookingCalendarWriter.php [hunk 1]
```diff
@@ -236,4 +253,89 @@ public function write(AppointmentConfig $config, DateTimeImmutable $start, strin
...
+		$vcalendar = new VCalendar();
+		$vtimezone = $vcalendar->createComponent('VTIMEZONE');
+		$vtimezone->TZID = $timezone;
+
+		$standard = $daylightStart = null;
+		foreach ($transitions as $i => $trans) {
+			$component = null;
+
+			// skip the first entry...
+			if ($i == 0) {
```
Comment (ChristophWurst) [inline lib/Service/Appointments/BookingCalendarWriter.php:291]:
```suggestion
			if ($i === 0) {
```

it's copied code but let's bring it up to our standards

**Review by ChristophWurst — COMMENTED** [review]
Could we have a test or two for the new timezone generation?

**Comment by ChristophWurst** [comment]
It could make sense to create a new class for the tz generation. That makes testing easier and removes complexity from the appointment code.

File: lib/Service/Appointments/TimezoneGenerator.php [hunk 2]
```diff
@@ -0,0 +1,150 @@
+<?php
+/*
+ * *
+ *  * {$app} App
```
Comment (ChristophWurst) [inline lib/Service/Appointments/TimezoneGenerator.php:4]: weird comment

File: lib/Service/Appointments/TimezoneGenerator.php [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+ *  *
+ *  * @copyright 2022 Anna Larch <anna.larch@gmx.net>
```
Comment (ChristophWurst) [inline lib/Service/Appointments/TimezoneGenerator.php:6]: happy new year

**Comment by ChristophWurst** [comment]
The generator looks great! Let's add some unit tests :)

**Comment by toabi** [comment]
Hi, what's the status about this? We have lots of issues with managing events with Customers using outlook… they don't really read text and just trust their Outlook Events. Which are currently offset by 1/2 hours… probably because of this bug.

**Comment by miaulalala** [comment]
This needs some tests, then we're good to merge. I hope to get this in before the next release on thursday.

File: lib/Service/Appointments/TimezoneGenerator.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCA\Calendar\Service\Appointments;
+
+use Sabre\VObject\Component;
+use Sabre\VObject\Component\VCalendar;
+use Sabre\VObject\TimeZoneUtil;
```
Comment (ChristophWurst) [inline lib/Service/Appointments/TimezoneGenerator.php:53]: This will contribute to https://github.com/nextcloud/calendar/issues/3728

**Review by ChristophWurst — APPROVED** [review]
Code looks good. Extraneous classes will byte us when they change

File: lib/Service/Appointments/TimezoneGenerator.php (hunk 4 above)
Comment (miaulalala) [reply lib/Service/Appointments/TimezoneGenerator.php:53]: I know. Hard to do this right now, but the refactoring is on my to do.

File: lib/Service/Appointments/TimezoneGenerator.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+/*
+ * @copyright 2023 Anna Larch <anna.larch@gmx.net>
+ *
+ * @author 2023 Anna Larch <anna.larch@gmx.net>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
```
Comment (ChristophWurst) [inline lib/Service/Appointments/TimezoneGenerator.php:36]: is this what they call dual licensing?

**Review by ChristophWurst — APPROVED** [review]

File: lib/Service/Appointments/TimezoneGenerator.php (hunk 5 above)
Comment (kesselb) [reply lib/Service/Appointments/TimezoneGenerator.php:36]: :rofl:

**Pull request merged**